packages/connect/src/handlers.js

```diff
@@ -164,9 +164,12 @@ function set(target, key, value, receiver) {
   }
 
   // If both the old value and the new value are objects OR if both are arrays, deepOverwrite.
+  // TODO: this should probably be refactored
   if (
-    isObject(oldValue) === isObject(value) ||
-    Array.isArray(oldValue) === Array.isArray(value)
+    !isPrimitive(oldValue) &&
+    !isPrimitive(value) &&
```

Comment:
Why do we need this?

Reply:
I think this should make it a bit clearer:
A and B are the old and new value respectively.

| A  \ B   | primitive | array    | object   |
| ------- | -------   | ------- | --------|   
|primitive|   reassign|  reassign| reassign|
| array  |   reassign |  **deepOverwrite** |reassign|
| object  |   reassign|  reassign|**deepOverwrite**|